host/sgx/loadelf.c

```diff
@@ -834,6 +1025,161 @@ static oe_result_t _sgx_update_enclave_properties(
     return result;
 }
 
+static oe_result_t _find_module_image(
+    const char* enclave_path,
+    oe_enclave_image_t* image,
+    char** module_path)
+{
+    oe_result_t result = OE_OK;
+    elf64_dyn_t* section_data = NULL;
+    size_t section_size = 0;
+    uint64_t number_of_entries;
+    elf64_addr_t strtab_offset = 0;
+    elf64_xword_t needed_offset = 0;
+    char* module_name = NULL;
+    char* path_name = NULL;
+    size_t path_size = 0;
+    const char* n = NULL;
+    const char* p = NULL;
+
+    assert(enclave_path);
+    assert(image);
+
+    if (module_path)
+        *module_path = NULL;
+
+    if (elf64_find_section(
+            &image->elf.elf,
+            ".dynamic",
+            (uint8_t**)&section_data,
+            &section_size) != 0)
+        goto done;
+
+    if (!section_data || !section_size)
+        OE_RAISE(OE_UNEXPECTED);
+
+    number_of_entries = section_size / sizeof(elf64_dyn_t);
+    for (uint64_t i = 0; i < number_of_entries; i++)
+    {
+        /* Explicitly prevent the use of DT_RPATH and DT_RUNPATH that affects
+         * the enclave measurement */
+        if (section_data[i].d_tag == DT_RPATH ||
+            section_data[i].d_tag == DT_RUNPATH)
+            OE_RAISE_MSG(
+                OE_UNEXPECTED,
+                "RPATH or RUNPATH should not be used in the enclave binary",
+                NULL);
+
+        if (section_data[i].d_tag == DT_STRTAB)
+            strtab_offset = section_data[i].d_un.d_ptr;
+        else if (section_data[i].d_tag == DT_NEEDED)
+            needed_offset = section_data[i].d_un.d_val;
+    }
+
+    /* Early return if either of the offset is not set (both are required to
+     * locate the module name) */
+    if (!strtab_offset || !needed_offset)
+        goto done;
+
+    module_name =
+        (char*)(image->elf.image_base + strtab_offset + needed_offset);
+    if (!module_name)
+        OE_RAISE(OE_UNEXPECTED);
+
+    /* Extract the module name from the path */
+    n = module_name + strlen(module_name);
+    while ((n - 1) >= module_name && *(n - 1) != '/' && *(n - 1) != '\\')
+        --n;
+
+    /* Find out the folder from the enclave path */
+    p = enclave_path + strlen(enclave_path);
+    while (p != enclave_path && *p != '/' && *p != '\\')
+        --p;
+
+    /* Allocate string to hold the module path */
+    path_size = (size_t)(p - enclave_path) + strlen(n) + 2;
+    path_name = calloc(1, path_size);
+    if (!path_name)
+        OE_RAISE(OE_OUT_OF_MEMORY);
+
+    if (p != enclave_path)
+        snprintf(
+            path_name,
+            path_size,
+            "%.*s/%s",
+            (int)(p - enclave_path),
+            enclave_path,
+            n);
+    else /* Handle the case if the module path does not include directories */
+        snprintf(path_name, path_size, "%s", n);
+
+    *module_path = path_name;
+
+done:
+    return result;
+}
+
+static oe_result_t _load_module_image(
```

Comment:
Again, consider renaming to more descriptive names e.g. `_load_dependency_image`, `_find_dependency_image` etc.

Reply:
Updated.